### PR TITLE
Batch aggregation

### PR DESCRIPTION
**Pull request by zwolf**
Adds the necessary pieces to implement batch aggregation. In total, they should include:

Aggregation table migration
Client to connect to aggregation service
Refactors of: model, factory, controller, schemas, policy, serializer
Specs of all the above

This is ready for review.

# Review checklist

- [ ] First, the most important one: is this PR small enough that you can actually review it? Feel free to just reject a branch if the changes are hard to review due to the length of the diff.
- [ ] If there are any migrations, will they the previous version of the app work correctly after they've been run (e.g. the don't remove columns still known about by ActiveRecord).
- [ ] If anything changed with regards to the public API, are those changes also documented in the `apiary.apib` file?
- [ ] Are all the changes covered by tests? Think about any possible edge cases that might be left untested.

### CODE REVIEW

File: spec/controllers/api/v1/aggregations_controller_spec.rb [hunk 1]
```diff
@@ -1,169 +1,87 @@
...
+              user: authorized_user.id.to_s,
               workflow: workflow.id.to_s
             }
           }
       }
     end
 
-    it_behaves_like "is creatable"
+    before do
+      allow(AggregationClient).to receive(:new).and_return(mock_agg)
```
Comment (yuenmichelle1) [inline spec/controllers/api/v1/aggregations_controller_spec.rb:50]: Not blocking, but maybe we could add a test where calling out with AggregationClient errors that ensures we receive the correct service unavailable response?

**Review by yuenmichelle1 — APPROVED** [review]
Looks fine. I had a couple non blocking comments and questions. 
Since this feature has pretty much been siloed to just you, Cliff, (maybe Haley and Coleman?). 
I think its worth adding more docs or comments on the PR so any one of us can go back and save some time on getting context. 


Some things:
- I think its worth documenting the flow of this to aggregations and back. Currently I'm not entirely sure of what the difference is between a task id vs the uuid vs the aggregations id. 
- Other thing I am wondering about, from what I understand of the new aggregations model, it is 1 aggregation per user per workflow. 
      - What happens if a user wants to re-run an aggregation for a workflow? (Does an aggregation record on panoptes get "cleaned" up after a certain point? Do we manually have to remove the aggregation record from the db so that re-run can be done? Is it even a thing to "re-run"?)

File: app/schemas/aggregation_create_schema.rb [hunk 2]
```diff
@@ -1,26 +1,35 @@
...
-      type "object"
-      additional_properties false
+    property 'task_id' do
+      type 'string'
+    end
 
-      required "subject", "workflow"
+    property 'status' do
+      type 'string'
+    end
```
[the remark this answers is not in the document]
Comment (zwolf) [reply app/schemas/aggregation_create_schema.rb:18]: I'll add the descriptions to a new docs page.

File: spec/policies/aggregation_policy_spec.rb [hunk 3]
```diff
@@ -97,7 +87,7 @@
         let(:api_user) { ApiUser.new(admin_user, admin: true) }
 
         it 'includes everything' do
```
[the remark this answers is not in the document]
Comment (zwolf) [reply spec/policies/aggregation_policy_spec.rb:89]: The difference is the context, the api_user is an admin but is not the resource owner. I'll change this from "includes everything" to "includes aggregation" to make it more clear that there aren't more, they should just be allowed.

File: spec/policies/aggregation_policy_spec.rb [hunk 4]
```diff
@@ -24,28 +22,24 @@
...
         let(:api_user) { ApiUser.new(logged_in_user) }
 
-        it "includes aggregations from public projects" do
-          expect(resolved_scope).to match_array(public_aggregation)
+        it 'returns nothing' do
+          expect(resolved_scope).to be_empty
         end
       end
 
       context 'for the resource owner' do
```
[the remark this answers is not in the document]
Comment (zwolf) [reply spec/policies/aggregation_policy_spec.rb:38]: Yes, and sure thing

File: spec/policies/aggregation_policy_spec.rb [hunk 5]
```diff
@@ -24,28 +22,24 @@
...
         let(:api_user) { ApiUser.new(anonymous_user) }
 
-        it "includes aggregations from public projects" do
-          expect(resolved_scope).to match_array(public_aggregation)
+        it 'returns nothing' do
+          expect(resolved_scope).to be_empty
         end
       end
 
       context 'for a normal user' do
```
[the remark this answers is not in the document]
Comment (zwolf) [reply spec/policies/aggregation_policy_spec.rb:30]: Yeah, just a different user that is logged in but has no permissions via the policy.

File: app/models/aggregation.rb [hunk 6]
```diff
@@ -1,20 +1,19 @@
...
-  belongs_to :subject
+  belongs_to :user
+  validates :workflow, presence: true
+  validates :user, presence: true
+  validates :user_id, uniqueness: { scope: :workflow_id }
 
-  validates_presence_of :workflow, :subject, :aggregation
-  validates_uniqueness_of :subject_id, scope: :workflow_id
-  validate :aggregation, :workflow_version_present
+  self.ignored_columns = %w[subject_id aggregation]
```
[the remark this answers is not in the document]
Comment (zwolf) [reply app/models/aggregation.rb:12]: This was originally to appease strong migrations, but it looks like the migration will run without it. I'll remove it and push and make sure it doesn't complain again.

File: spec/controllers/api/v1/aggregations_controller_spec.rb (hunk 1 above)
Comment (zwolf) [reply spec/controllers/api/v1/aggregations_controller_spec.rb:50]: ✔️

File: spec/controllers/api/v1/aggregations_controller_spec.rb [hunk 7]
```diff
@@ -1,169 +1,114 @@
...
+      expect(Aggregation.first.task_id).to eq('asdf-1234-asdf')
+    end
 
-        it "should return not_found" do
-          get :show, params: { id: resource.id }
-          expect(response).to have_http_status(:not_found)
-        end
-      end
+    context 'when there is an existing aggregation for that user and workflow' do
+      let!(:existing_agg) { create(:aggregation, workflow: workflow, user: authorized_user) }
```
[the remark this answers is not in the document]
Comment (zwolf) [reply spec/controllers/api/v1/aggregations_controller_spec.rb:72]: needs to exist in the db to force collision

**Comment by zwolf** [comment]
I had some trouble with Doorkeeper scopes when I was testing this API outside of the specs (not Pundit scopes, but [Doorkeeper's own](https://doorkeeper.gitbook.io/guides/configuration/scopes)). Nothing else in the app scopes solely by workflow and just adding to the [optional scopes](https://github.com/zooniverse/panoptes/blob/master/config/initializers/doorkeeper.rb#L5) wasn't enough. Rather than keep fighting it, I switched the base scope for the Aggregation resource back to Project and added that column to the table--that's all it's used for so far, but the relation is logical and Doorkeeper is happy.

> I think its worth documenting the flow of this to aggregations and back.

Added a page to the docs for the new model. @lcjohnso I think you wanted a ping to have a look.

>Semi related to ^, I'm currently not entirely sure of what the difference is between a task id vs the uuid vs the aggregations id. Maybe adding a description on the schema or more description on PR would be helpful?

Included this info in the new docs. The task_id comes from Celery via Aggregation and represents the running background task. This id can be used to check on the status of a running job via the aggregation API and to see if it's been completed. This could be implemented by the front end as a way to check the status of pending, long-running jobs. The UUID is generated by the aggregation run and used in filenames and folders on blob storage. It'll be used to construct a URL that the files will be downloaded directly from.

> Other thing/s I am wondering about: From what I understand of the new aggregations model, it is 1 aggregation per user per workflow. What happens if a user wants to re-run an aggregation for a workflow? (Does an aggregation record on panoptes get "cleaned" up after a certain point? Do we manually have to remove the aggregation record from the db so that re-run can be done? Is it even a thing to "re-run"?)

Yes, aggregations are unique per user & workflow. And true, this was missing a way to re-run the aggregation--what makes the most sense? The FE checks for an existing aggregation resource via `GET api/aggregations?workflow_id=[wf_id]&user_id=[user_id]`. I have now implemented DELETE on the controller and added a couple specs to more clearly indicate what the intended behavior is. And though I was trying to avoid making the FE worry about it, the standard approach might be easiest: FE modal checks for existing agg, and if it's completed/failed, a click of "rerun" DELETEs by id, then POSTs to create new one, triggering a new aggregation run. 

Talk actually does clean up data requests after a day and blocks early rerun attempts, I don't love that either. For now, I think this functionality is enough to work with, but maybe I could add some kind of shortcut (via route, POST body short circuit, other?) to reduce the required FE calls. Thoughts?

File: spec/policies/aggregation_policy_spec.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (zwolf) [reply spec/policies/aggregation_policy_spec.rb]: Clarified in specs

**Comment by lcjohnso** [comment]
Docs look good -- thanks!

One question regarding the existing aggregations table: I see you're removing the main `aggregation` data column in the migration, but will the >1 million rows stay in place? Or do those need to be cleaned up too? Is that happening elsewhere, or in this or a similar migration?

**Comment by zwolf** [comment]
This is ready to merge. I truncated the `aggregations` table on staging in preparation:

```ruby
irb(main):005:0> Aggregation.count
[primary]   (55.2ms)  SELECT COUNT(*) FROM "aggregations"
=> 1
irb(main):007:0> Aggregation.connection.truncate(Aggregation.table_name)
[primary]   (777.3ms)  TRUNCATE TABLE "aggregations"
=> #<PG::Result:0x00007fad5f56b968 status=PGRES_COMMAND_OK ntuples=0 nfields=0 cmd_tuples=0>
irb(main):008:0> Aggregation.count
[primary]   (65.2ms)  SELECT COUNT(*) FROM "aggregations"
=> 0
```

I plan on doing the same thing to production before this gets deployed. In the end, this feels like a data operation rather than a schema change and not something that should exist as a migration or a repeatable rake task. Truncate is the fastest way to remove the data, which will be important on prod where there are ~1.2M row (newest from 2017).

**Pull request merged**